### DISCUSSION TITLE
Key derivation failed

### DISCUSSION THREAD

**Discussion by keerthanakalyan** [strongswan/strongswan#2320] [Q&A]
Hello all,
We are seeing a key derivation failure when FIPS is enabled in openssl and strongswan.
OS: Photon OS
strongSwan version: 5.9.8 

charon 2288 - - 07[CFG] selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_256
charon 2288 - - 07[IKE] key derivation failed

ike and esp configurations used

  ike=aes128-sha256-ecp256
  esp=aes128gcm8-sha256-modp2048

When we disable FIPS, this issue is not seen.

Thanks in advance.

**Comment by tobiasbrunner** [comment]
Which OpenSSL version are you using?

**Reply by keerthanakalyan** [reply]
We are using FIPS certified openssl version 3.0.8

**Reply by tobiasbrunner** [reply]
The key derivation uses OpenSSL's HKDF implementation, which is available since 1.1.1 and I'd assume it would not get disabled due to FIPS mode. I also can't reproduce this when running the test vectors against a FIPS-enabled OpenSSL 3.0.8, only the expected vectors fail:

    ...
    enabled  KDF_PRF[openssl]: passed 6 test vectors
    enabled  KDF_PRF_PLUS[openssl]: passed 7 test vectors
    ...
    Failure in 'test_vectors': success: test vector for CAMELLIA_CBC from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 0)
    Failure in 'test_vectors': success: test vector for CAMELLIA_CBC from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 0)
    Failure in 'test_vectors': success: test vector for CAMELLIA_CBC from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 0)
    Failure in 'test_vectors': success: test vector for CHACHA20_POLY1305 from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 1)
    Failure in 'test_vectors': success: test vector for HMAC_MD5_96 from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 2)
    Failure in 'test_vectors': success: test vector for HMAC_MD5_128 from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 2)
    Failure in 'test_vectors': success: test vector for HASH_MD4 from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 3)
    Failure in 'test_vectors': success: test vector for HASH_MD5 from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 3)
    Failure in 'test_vectors': success: test vector for PRF_HMAC_MD5 from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 4)
    Failure in 'test_vectors': success: test vector for MODP_2048_224 from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 9)
    Failure in 'test_vectors': success: test vector for MODP_2048_256 from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 9)
    Failure in 'test_vectors': success: test vector for MODP_1536 from 'openssl' plugin failed (../../../../strongswan/src/libstrongswan/tests/suites/test_vectors.c:50, i = 9)

**Reply by keerthanakalyan** [reply]
Thank you very much for the response. 

It looks like fips issue was not the actual issue, it has do with openssl version 3.0.14.
We upgraded our openssl from 3.0.13 to 3.0.14 and we started noticing issues with the tunnels. FIPS was a diversion. Though disabling FIPS showed ipsec status all fine, but the ping between the tunnels does not work.

Is strongswan version 5.9.8 with openssl plugins compatible with openssl version 3.0.14?

**Reply by tobiasbrunner** [reply]
Why shouldn't it be compatible? Also, why would OpenSSL be involved in handling IPsec traffic at all? That's usually handled by the kernel, which doesn't use OpenSSL. Or are you using [kernel-libipsec](https://docs.strongswan.org/docs/5.9/plugins/kernel-libipsec.html) for some reason?

**Comment by keerthanakalyan** [comment]
We could root cause the issue. Issue was because of this regression in openssl with older provider. 
https://github.com/openssl/openssl/issues/24611

Thank you @tobiasbrunner for your response and suggestions.

**Reply by tobiasbrunner** [reply]
Good to know, thanks for the info.
